File: app/src/main/java/de/karzek/diettracker/presentation/main/cookbook/adapter/viewHolder/RecipeCookbookSearchResultViewHolder.java
package de.karzek.diettracker.presentation.main.cookbook.adapter.viewHolder;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.TextView;

import com.daimajia.swipe.SwipeLayout;

import java.io.ByteArrayOutputStream;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import de.hdodenhof.circleimageview.CircleImageView;
import de.karzek.diettracker.R;
import de.karzek.diettracker.presentation.model.GroceryDisplayModel;
import de.karzek.diettracker.presentation.model.RecipeDisplayModel;

/**
 * Created by MarjanaKarzek on 30.05.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 30.05.2018
 */
public class RecipeCookbookSearchResultViewHolder extends RecyclerView.ViewHolder {

    @BindView(R.id.recipe_image)
    CircleImageView recipeImage;
    @BindView(R.id.recipe_title)
    TextView recipeTitle;
    @BindView(R.id.swipe_layout)
    SwipeLayout swipeLayout;

    private final OnRecipeItemClickedListener onRecipeItemClickedListener;
    private final OnRecipeAddPortionClickedListener onRecipeAddPortionClickedListener;
    private final OnRecipeEditClickedListener onRecipeEditClickedListener;
    private final OnRecipeDeleteClickedListener onRecipeDeleteClickedListener;

    public RecipeCookbookSearchResultViewHolder(ViewGroup viewGroup,
                                                OnRecipeItemClickedListener onRecipeItemClickedListener,
                                                OnRecipeAddPortionClickedListener onRecipeAddPortionClickedListener,
                                                OnRecipeEditClickedListener onRecipeEditClickedListener,
                                                OnRecipeDeleteClickedListener onRecipeDeleteClickedListener) {
        super(LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.item_cookbook_recipe, viewGroup, false));
        ButterKnife.bind(this, itemView);

        this.onRecipeItemClickedListener = onRecipeItemClickedListener;
        this.onRecipeAddPortionClickedListener = onRecipeAddPortionClickedListener;
        this.onRecipeEditClickedListener = onRecipeEditClickedListener;
        this.onRecipeDeleteClickedListener = onRecipeDeleteClickedListener;
    }

    public void bind(RecipeDisplayModel recipe) {
        if (recipe.getPhoto() != null)
            recipeImage.setImageBitmap(BitmapFactory.decodeByteArray(recipe.getPhoto(), 0, recipe.getPhoto().length));
        recipeTitle.setText(recipe.getTitle());
        itemView.setTag(recipe.getId());
    }

    @OnClick(R.id.item_layout)
    public void onItemClicked() {
        swipeLayout.close();
        onRecipeItemClickedListener.onItemClicked((int) itemView.getTag());
    }

    @OnClick(R.id.swipe_option_portion)
    public void onAddPortionClicked() {
        onRecipeAddPortionClickedListener.onAddPortionClicked((int) itemView.getTag());
    }

    @OnClick(R.id.swipe_option_edit)
    public void onEditRecipeClicked() {
        swipeLayout.close();
        onRecipeEditClickedListener.onEditRecipeClicked((int) itemView.getTag());
    }

    @OnClick(R.id.swipe_option_delete)
    public void onDeleteRecipeClicked() {
        swipeLayout.close();
        onRecipeDeleteClickedListener.onDeleteRecipeClicked((int) itemView.getTag());
    }

    public interface OnRecipeItemClickedListener {
        void onItemClicked(int id);
    }

    public interface OnRecipeAddPortionClickedListener {
        void onAddPortionClicked(int id);
    }

    public interface OnRecipeEditClickedListener {
        void onEditRecipeClicked(int id);
    }

    public interface OnRecipeDeleteClickedListener {
        void onDeleteRecipeClicked(int id);
    }

}

File: app/src/main/java/de/karzek/diettracker/presentation/main/cookbook/adapter/viewHolder/RecipeCookbookSearchResultViewHolder.kt
package de.karzek.diettracker.presentation.main.cookbook.adapter.viewHolder

import android.graphics.BitmapFactory
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import de.karzek.diettracker.presentation.model.RecipeDisplayModel
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.item_cookbook_recipe.*

/**
 * Created by MarjanaKarzek on 30.05.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 30.05.2018
 */
class RecipeCookbookSearchResultViewHolder(override val containerView: View,
                                           private val onRecipeItemClickedListener: OnRecipeItemClickedListener,
                                           private val onRecipeAddPortionClickedListener: OnRecipeAddPortionClickedListener,
                                           private val onRecipeEditClickedListener: OnRecipeEditClickedListener,
                                           private val onRecipeDeleteClickedListener: OnRecipeDeleteClickedListener) : RecyclerView.ViewHolder(containerView), LayoutContainer {

    fun bind(recipe: RecipeDisplayModel) {
        if (recipe.photo != null)
            recipe_image.setImageBitmap(BitmapFactory.decodeByteArray(recipe.photo, 0, recipe.photo.size))
        recipe_title.text = recipe.title
        itemView.tag = recipe.id

        setupClickListeners()
    }

    private fun setupClickListeners() {
        item_layout.setOnClickListener {
            swipe_layout.close()
            onRecipeItemClickedListener.onItemClicked(itemView.tag as Int)
        }
        swipe_option_portion.setOnClickListener { onRecipeAddPortionClickedListener.onAddPortionClicked(itemView.tag as Int) }
        swipe_option_edit.setOnClickListener {
            swipe_layout.close()
            onRecipeEditClickedListener.onEditRecipeClicked(itemView.tag as Int)
        }
        swipe_option_delete.setOnClickListener {
            swipe_layout.close()
            onRecipeDeleteClickedListener.onDeleteRecipeClicked(itemView.tag as Int)
        }
    }

    interface OnRecipeItemClickedListener {
        fun onItemClicked(id: Int)
    }

    interface OnRecipeAddPortionClickedListener {
        fun onAddPortionClicked(id: Int)
    }

    interface OnRecipeEditClickedListener {
        fun onEditRecipeClicked(id: Int)
    }

    interface OnRecipeDeleteClickedListener {
        fun onDeleteRecipeClicked(id: Int)
    }

}
